processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -304,31 +306,79 @@ private void handleDefinedNestedTargetMapping(Set<String> handledTargets) {
 
                 Map<Parameter, MappingOptions> optionsBySourceParam = entryByTP.getValue().groupBySourceParameter();
                 boolean forceUpdateMethod = optionsBySourceParam.keySet().size() > 1;
+                PropertyEntry targetProperty = entryByTP.getKey();
                 for ( Entry<Parameter, MappingOptions> entryByParam : optionsBySourceParam.entrySet() ) {
```

Comment:
I think that we need to extract all this loops and creation of mapping options into one method / class instead of doing it one by one. And most importantly I think that we have to add some unit tests only for that. I think that the entire logic here is quite complex.